packages/xstate-react/src/useActor.ts

```diff
@@ -1,13 +1,29 @@
 import { useState, useRef } from 'react';
 import useIsomorphicLayoutEffect from 'use-isomorphic-layout-effect';
-import { Sender, ActorRefLike } from './types';
-import { EventObject, Actor } from 'xstate';
+import { Sender } from './types';
+import { ActorRef, EventObject } from 'xstate';
 import useConstant from './useConstant';
 
-export function useActor<TEvent extends EventObject, TEmitted = any>(
-  actorRef: ActorRefLike<TEvent, TEmitted> | Actor,
-  getSnapshot: (actor: typeof actorRef) => TEmitted = (a) =>
-    'state' in a ? a.state : (undefined as any)
+function isActorWithState<T extends ActorRef<any>>(
+  actorRef: T
+): actorRef is T & { state: any } {
+  return 'state' in actorRef;
+}
+
+function isDeferredActor<T extends ActorRef<any>>(
+  actorRef: T
+): actorRef is T & { deferred: boolean } {
+  return 'deferred' in actorRef;
+}
+
+export function useActor<
```

Comment:
I have no idea why yet but this signature is being problematic with `strictFunctionTypes` setting:
https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&alwaysStrict=false&noEmitHelpers=true#code/JYOwLgpgTgZghgYwgAgPICMDO0Bu0A8AKgHzIDeAUMsiBAB5gBcyAFDnADYCuEzhAlMgC8pHAHtgAEwDcVZNChiozFguZwQAT0Ejk4qbOoIxAWwAOHCJBU7REmRQC+FCqEixEKAMpcsCKMBmYMBiIORyXCCYvpj+wOgQLPzM+g7OruDQ8EjIPn4B6HDolkSklNTR+fGJYli40MwY2FB4UKXJuTFxQSEghsiVsQWJctS0DCrs3LzIAsJ2UgA0o-JQilAA-CoKSupatnr2y9RGphZWEFusB6lyHXlDgcGhsuluWZ7IAIIIYEoAShAYEQAKJ4cDyBgQECSTDkZBgTRmGaYMABEAAc2QjkWsxBJmAYEgkmEyA0mmIcnokBhcIecUKxQgoIJRIgkjKcmwMO24KYeL5N3srxciORyDB0LAGAAVhBfqSyGKUWjQBj0pJ5Rw4FAUDBIr9esguNgfn82nJCJKIdTobCJXzZfKwMc8azia7CGalJCafbvVBAcCrXzcVb3ezSEJvr8AUDQaG3YTiZTiCw5IhzUG+AHXRirF4QHAzJgABZiMBXFiZ3azAMHcPJ9nICgAelbJ2j1dsbY7J2QwEwAYA6oTS14wHBIN3kBsyQA6VFTlDMSKamCgdkUDoAbUbbMkuNUfL41rAQqkAF0XJqENrdchjFEwGS4HBmAGg-gyAikTMAOQwGIYj-tiuIgFwJgJFAlIUE+qLIDuS6QLi3KSJepImhAAbVm+-BAA

The only thing I've recognized so far about this problem is that introducing `TActor` generic is causing this.

Reply:
Ok... I **think** I might somewhat understand this now. Please take a look at this even more simplified demo:
[TS playground](https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&alwaysStrict=false&noEmitHelpers=true#code/JYOwLgpgTgZghgYwgAgIILAeygJQjAHgChlkAVAUQDcJxkIAPSEAEwGdkBvZMATwAcIALmRswUUAHNkAXwA0RAHxcSo2ixEAKCDXAjKusAEpkAXmVVMwFgG4iMokRYQEAGzhQUMAK4gMwTBBkbzYIdCwoYlIDWjB6JnUObj5BETEJEGl5VTJw7HjmdjQMbDxCGPBFJU1VRAiy-TyoIiMRS2tHZzcPFARAsWQ67BEmsoJkgWFkAHIYTExp2SqiELCSqE0hqCMV0KbxnkmRWfnF+WL6-AOUqZOFpcVN9aMgA)

We may notice that the inferred generics for here are as follows:
```ts
function useActor<{
    type: string;
}, ActorRef<{
    type: string;
}>>(actorRef: ActorRef<{
    type: string;
}>): void
```
And this is the same with or without `strictFunctionTypes` setting. For some reason, TS infers here the least strict type for `TEvent` that satisfies this. So it's an inference limitation because if you take a look at the provided demo there is no problem is we supply generic explicitly.

Why this is a problem even without `strictFunctionTypes`? Well, the inferred and returned `Sender<TEvent>` is too loose so it allows incorrect calls:
[TS playground](https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&alwaysStrict=false&noEmitHelpers=true#code/JYOwLgpgTgZghgYwgAgIILAeygJQjAHgChlkAVAUQDcJxkIAPSEAEwGdkBvZMATwAcIALmRswUUAHNkAXwA0RAHxcSo2ixEAKCDXAjKusAEpkAXmVVMwFgG4iMokRYQEAGzhQUMAK4gMwTBBkbzYIdCwoYlIDWjB6JnUObj5BETEJEGl5VTJw7HjmdjQMbDxCGPBFJU1VRAiy-TyoIiMtHX1qWJNzZEtrR2c3DxQEQLFkOuwRJrKCZIFhZAByGExMJdkqolGQcdDWAEYzYNCmzUmoI0d9lgPNedTlgHUACVRKADUKHA2ZK+2xnEbgAmY4hMIlSIPRYrNa-OTFer4OY8BYiWHrTaKc6QoxAA)

Why that code errors with `strictFunctionTypes` setting? Because of its sole nature that changes what function types can be assigned to what function types (variance-related) - it's explained in [this SO answer](https://stackoverflow.com/a/51767568) and showcased in the corresponding demo:
[TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAEghgOwCYBsICcoF4oAp0QCOArhAM7ABcUASkaRQJTYB8t5YA9gmRAFB8AxtwpQAtiHjI06alNQZseAiXJV2q0QDIoAbygQAHsHRwAgugDm1CugCWCS1AC+zLG118o3qCobAAOiMTcysA4E4AVTBIdABhOF5cRgBuLx8CYGJ0BCgEYhQUKETihBA05yA)

I've also tried to play around with our case on some even more simplified case and to my surprise, I haven't replicated the behavior 1 to 1. We indeed get the same problem if we include the function property (and the generic isn't inferred to the most narrow type possible):
[TS playground](https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false#code/C4TwDgpgBAQg9gDwDwGUoQcCA7AJgZyn2ACcBLbAcwD4oBeKAbwCgooA3ALigApiTuKAJT1a7OGVzMAvs2a4IAYwA2AQxLQAZgFdsi4GTjYo2-BBSkKlVOkw4CRS1QA0sW1jyF4yFNWo8AI0RuGCFucUk5BRV1aEUjYigghBDEJAByTTg4dOpmU3MnSkDEISA)
but if we make this a regular property a weird thing happens to inferred generics:
[TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQg9gDwDwGUoQcCA7AJgZyn2ACcBLbAcwD4oBeKAbwCgooA3ALihWYF9mzXBADGAGwCGJaADMArthHAycbFDn4IKUhUqp0mHASI6qAGlgGseQvGQpq1ABQAjRNxgBKbuzhlcgsLiUtAiqsRQbggeiEgA5DJwcHHUzBpappSuiJ5AA)
By inspecting the `useString` call we can see this:
```ts
function useString<string, Box<"foo">>(box: Box<"foo">): void
```
which has a different type inferred at 2 different positions even though it should be the same according to the declaration 😱 

---

Either way - it seems that we just can't rely on inference here. It's just incapable of inferring this properly right now. I propose to make the second argument a regular value that has be given upfront - and if it's not then we can try to access `actorRef.state`. This should simplify the types and the inference.

Reply:
In that case, I think it might be better to have the full actor as the first generic type param:

```ts
declare function useActor<
  TActor extends ActorRef<any, any> = ActorRef<any, any>
>(
  actorRef: TActor,
  getSnapshot?: (actor: TActor) => TActor extends ActorRef<any, infer TEmitted> ? TEmitted : never
): [TActor extends ActorRef<any, infer TEmitted> ? TEmitted : never, (event: TActor extends ActorRef<infer TEvent, any> ? TEvent : never) => void]

declare const aaa: ActorRef<{ type: 'foo' }, number>

const [state, send] = useActor(aaa)
```

[TS Playground](https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&alwaysStrict=false&noEmitHelpers=true#code/JYOwLgpgTgZghgYwgAgPICMDO0Bu0A8AKgHzIDeAUMsiBAB5gBcyAFDnADYCuEzhAlMgC8pHAHtgAEwDcVZNChiozFguZwQAT0Ejk4qbOoIxAWwAOHCJBU7REmRQC+FCqEixEKAMpcsCKMBmYMBiIORyXCCYvpj+wOgQLPzM+g7OruDQ8EjIPn4B6HDolkSklNTR+fGJYli40MwY2FB4UKXJuTFxQSEghsiVsQWJctS0DCrs3LzIAsJ2UgA0o-JQilAA-CoKSupatnr2y9RGphZWEFusB6lyHXlDgcGhsuluWZ7IAIIIYEoAShAYEQAKJ4cDyBgQECSTDkZBgTRmGaYMABEAAc2QjkWsxBJmAYEgkmEyA0mmIcnokBhcIecUKxQgoIJRIgkjKcmwMO24KYeL5N3srxciORyDB0LAGAAVhBfqSyGKUWjQBj0pJ5Rw4FAUDBIr9esguNgfn82nJCGalJCabDvr8AUD8OTceTSEIHebAcDXWStJTiCw5IhvUC+NaoMdkBirF4QHAzJgABZiMBXFih3azSMHK2OqC26H2yM+l1aXGgGDQPGs4mkDa1wnE5DMWitCgdADa+fNRdpXqdvoryCrNcI+Ob7IbTbZJLbEFauNUfIjBf7JYLZbHhYnfLdAeQjb3UtbNEX0CFUgAui5NQhtbrkMYomAyXA4MxS86yAikTMAHIYDEMQAOxXEQC4EwEigSkKBfVFkC7VE4EgXFuUka9SRNCBI0zD9+CAA)

> I propose to make the second argument a regular value that has be given upfront - and if it's not then we can try to access `actorRef.state`.

No, we shouldn't assume that `actorRef.state` exists; remember, you didn't want `ActorRef` to be too loose re: `{ send: ..., [key: string]: any }`.

Reply:
> In that case, I think it might be better to have the full actor as the first generic type param:

You could simplify this a little bit by using a single conditional type for the return type and I don't think the default for this generic changes anything here (they usually only make sense if you are having more than one type param) 🤔 

[TS Playground](https://www.typescriptlang.org/play?noEmitHelpers=true#code/JYOwLgpgTgZghgYwgAgPICMDO0Bu0A8AKgHzIDeAUMsiBAB5gBcyAFDnADYCuEzhAlMgC8pHAHtgAEwDcVZNChiozFguZwQAT0Ejk4qbOoIxAWwAOHCJBU7REmRQC+FCqEixEKAMpcsCKMBmYMBiIORyXCCYvpj+wOgQLPzM+g7OruDQ8EjIPn4B6HDolkSklNTR+fGJYli40MwY2FB4UKXJuTFxQSEghsiVsQWJctS0DCrs3LzIAsJ2UgA0o-JQilAA-CoKSupatnr2y9RGphZWEFusB6lyHXlDgcGhsuluWZ7IAIIIYEoAShAYEQAKJ4cDyBgQECSTDkZBgTRmGaYMABEAAc2QjkWsxBJmAYEgkmEyA0mmIcnokBhcIecUKxQgoIJRIgkjKcmwMO24KYeL5N3srxciORyDB0LAGAAVhBfqSyGKUWjQBj0pJ5Rw4FAUDBIr9esguNgfn82nJCGalJCabDvr8AUD8OTceTKcQWHJEObATA+NaoMdkBirF4QHAzJgABZiMBXFg+3azQMHK2OqC26H2wN+l1aXGgGDQPGs4mkDalwnE5DMWitCgddPmrO0h2+51FkuESXgQsgYuZntl9mU6iVgDaw+r7Nxqj5fF7YCFUgAunI6xAGxRNQhtbrkMYomAyXA4Mxc86yAikTMAOQwMRiO-Y3EgLgmBJQSkUI+o5ATqicCQLi3KSKupImhAgaJme-BAA)

> No, we shouldn't assume that actorRef.state exists; remember, you didn't want ActorRef to be too loose re: { send: ..., [key: string]: any }.

But we already assume that `.state` might exist at runtime - this ain't a completely type-safe assumption but it seems like a fair tradeoff here. I don't think the proposed solution would really be any different from what we have now:

```js
// current
function useActor(
  actorRef,
  getSnapshot = (a) => isActorWithState(a) ? a.state : undefined
) {
  // just for demo purposes
  const snapshot = getSnapshot(actorRef)
  // ...
}

// proposed
function useActor(
  actorRef,
  snapshot
) {
  // just for demo purposes
  snapshot = snapshot === undefined && isActorWithState(a) ? a.state  : undefined
  // ...
}
```

This would improve the TS inference while using the "standard" generics that we use across the codebase:
[TS playground](https://www.typescriptlang.org/play?noEmitHelpers=true#code/JYOwLgpgTgZghgYwgAgPICMDO0Bu0A8AKgHzIDeAUMsiBAB5gBcyAFDnADYCuEzhAlMgC8pHAHtgAEwDcVZNChiozFguZwQAT0Ejk4qbOoIxAWwAOHCJBU7REmRQC+FCqEixEKAMpcsCKMBmYMBiIORyXCCYvpj+wOgQLPzM+g7OruDQ8EjIPn4B6HDolkSklNTR+fGJYli40MwY2FB4UKXJuTFxQSEghsiVsQWJctS0DCrs3LzIAsJ2UgA0o-JQilAA-CoKSupatnr2y9RGphZWEFusB6lyHXlDgcGhsuluWZ7IAIIIYEoAShAYEQAKJ4cDyBgQECSTDkZBgTRmGaYMABEAAc2QjkWsxBJmAYEgkmEyA0mmIcnokBhcIecUKxQgoIJRIgkjKcmwMO24KYeL5N3srxciORyDB0LAGAAVhBfqSyGKUWjQBj0pJ5Rw4FAUDBIr9esguNgfn82nJCJKIdTobCJXzZfKwMc8azictiCw5IhzYCYMwzQCgaC+bire72cRXZgQHAzJgABZiMBXCOE4nICgdADa6bZklxqj5fGtYCFUgAui5NQhtbrkMYomAyb9dt821B-fgyAikTMAOQwMRiAfY3EgLgmBJQSkUJuo5A51FwSC47mSSukk0QINQFi+pT8IA)

There is ofc a little bit of ambiguity in this version regarding calls like `useActor(actor, undefined)` but it also seems like a fair tradeoff as this shouldn't affect any real-life code. I don't see any difference when it comes to runtime semantics between the current and proposed solutions - the only difference is how the snapshot value gets provided (getter vs explicitly).

I don't mind your solution much - just exploring alternatives here now that I believe are just somewhat simpler conceptually (the need to use 2 conditional types raises the complexity quite a bit when using a single `TActor` approach).

Reply:
I disagree with getting the snapshot value "explicitly" as a mutable value, as was done before. That exposes implementation details, and makes a wrongful assumption (that the snapshot is always available as a mutated property of the actor), and also limits flexibility with the `useActor` hook. Basically, this should behave like `getSnapshot` in the [`useMutableSource`](https://github.com/reactjs/rfcs/blob/master/text/0147-use-mutable-source.md) hook and Redux's `store.getState()`.

Reply:
Well, our usage of `getSnapshot` is different from the `useMutableSource` - they **always** read state through it, even when they are notified about a change through a subscription (Redux does the same as `useMutableSource`). And we only use it when initializing the state and in the subscribing effect.

I agree though that given the second way in which we are using it this has to be a getter function and can't be a plain value.

Reply:
> EDIT: also, the implementation might/will probably change in v5, re: getting the snapshot value. We should not hardcode reading from actor.state.

Agreed, although I'm not sure how to do that without requiring `getSnapshot` at all times and from what I remember (but maybe I'm wrong) you were opposed to "standardizing" a common property in the Actor interface that could be read for those purposes.

Reply:
> you were opposed to "standardizing" a common property in the Actor interface that could be read for those purposes.

Yes, I'm opposed to it because we should make actors a minimal `{ send, subscribe }` interface. The idea of having a property that "holds" a snapshot value is conceptually the same as having something that previously subscribed (or asked for via sending an event with a self-reference) to an actor and locally held (cached) the last-known-emitted value. But it shouldn't be a built-in part of actors.